enhancements/machine-api/control-plane-machinesets.md

```diff
@@ -0,0 +1,227 @@
+	- `machine.openshift.io/cluster-api-cluster":    clusterID`
+
+#### Declarative horizontal scaling
+- Each MachineSet can be scaled in/out by changing the replica number or via "scale" subresource.
+- Out of band the Cluster etcd Operator is responsible for watching the existing nodes and taking care of the etcd operational aspects.
+- During scaling in operations any machine deletion will always honour Pod Disruption Budgets (PBD). This gives etcd guard the chance to block a deletion that might be considered disruptive.
+
+#### Declarative Vertical scaling
+- With MachineSets in place this is semi-automated:
+	- A particular provider property can be changed by any consumer in the MachineSet spec.
+	- This won't affect existing Machines but it would apply for upcoming Machines.
+	- Machine creation with the new spec property can be forced by scaling out or deleting existing machines, e.g `kubectl delete machine name`
+
+#### Self healing (Machine Health Checking)
+- Specific MHC details can be found [here](https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/machine-health-checking.md)
+
+- The Installer will create a Machine Health Checking resource targeting the Control Plane Machines.
+
+```
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineHealthCheck
+metadata:
+  name: controlPlane 
+  namespace: openshift-machine-api
+spec:
+  selector:
+    matchLabels:
+      machine.openshift.io/cluster-api-machine-role: master
+      machine.openshift.io/cluster-api-machine-type: master
+  unhealthyConditions:
+  - type:    "Ready"
+    timeout: "300s"
+    status: "False"
+  - type:    "Ready"
+    timeout: "300s"
+    status: "Unknown"
+  maxUnhealthy: "34%"
+```
+
+-   Any machine deletion will always honour Pod Disruption Budgets (PBD). This gives etcd guard the chance to block a deletion that it considers disruptive.
+
+TODO: Currently multiple MHC can target the same Machine. We should revisit a stronger mechanisim to prevent undesired MHC to target Control Plane Machines.
+
+### Risks and Mitigations
+
+During horizontal scaling operations there are multiple sensitive scenarios like scaling from 1 to 2. As soon as the etcd API is notified of the new member the cluster loses quorum until that new member starts and joins the cluster. This as well as any other operational aspect of etcd healthiness must be still handled by the [Cluster etcd Operator](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/cluster-etcd-operator.md#motivation).
+The contract for the machine API is by honouring Pod Disruption Budgets (PBD), including those given by etcd guard.
+
+## Design Details
+
+### Test Plan
+
+- Exhaustive unit testing.
+- Exhaustive integration testing via [envTest](https://book.kubebuilder.io/reference/testing/envtest.html).
+- E2e testing on the [machine API](https://github.com/openshift/cluster-api-actuator-pkg/tree/master/pkg) and origin e2e test suite:
```

Comment:
Would like to see some tests when bad configuration is supplied.  I.e.
- scale in to less than 3 masters
- instance type too small to handle master workload

Reply:
At best we can test to block scale in on etcd guard PDB. For any farther automation that requires somehow auto scaling of compute this proposal is a necessary first step.